docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,203 @@
+
+`{ } >> { }`  
+  This is the descendant operator. The spans returned from this operator will match the right hand side conditions while also being descendants of spans that match the left hand side conditions.
+
+`{ } > { }`  
+  This is the child operator. The spans returned from this operator will match the right hand side conditions while also being children of spans that match the left hand side conditions.
+
+`{ } ~ { }`  
+  This is the sibling operator. The spans returned from this operator will match the right hand side conditions while also being siblings of spans that match the left hand side conditions.
+
+## Aggregators
+
+All of the above expressions involve asking questions about individual spans. However, sometimes we want to ask questions about a set of spans. For that we have aggregate functions.
+
+**Examples**
+
+Find traces where the total number of spans is greater than 10:  
+`count() > 10`
+
+Find traces where the average duration is greater than 1s:  
+`avg(duration) > 1s`
+
+## Expression Pipelining
+
+Pipelining allows us to "pipe" a set of spans from one expression to the next. This is particularly useful if you want to perform an aggregate over a subset of a trace.
+
+**Examples**
+
+Find traces that have more than 3 spans with an attribute `http.status` with a value of `200`:  
+`{ http.status = 200 } | count() > 3`
+
+## Grouping
+
+Grouping allows us to take a trace and break it down into sets of spans that are evaluated by future pipeline entries. Each set of spans created by the group is _individually_ evaluated by downstream expressions.
+
+**Examples**
+
+Find traces that have more than 5 spans in any region:  
+`by(region) | count() > 5 `
+
+## Examples
+
+Any span matches an attribute:  
+`{ namespace = "prod" }`
+
+Two attributes appear on the same span:  
+`{ namespace = "prod" && http.status = 200 }`
+
+Two attributes appear anywhere within a trace:  
+`{ namespace = "prod" } && { http.status = 200 }`
+
+Any span has a duration over one second:  
+`{ duration > 1s }`
+
+The trace as a whole has a duration of over one second:  
+`max(end) - min(start) > 1s`
+
+The average duration of spans exceeds 1s and any span has a specific attribute:
+`avg(duration) > 1s && { namespace = "prod" }`
+
+A trace has over 5 spans with http.status = 200 in any given namespace:  
+`{ http.status = 200 } | by(namespace) | count() > 5`
+
+A trace passed through two regions in a specific order:  
+`{ region = "eu-west-0" } >> { region = "eu-west-1" }`
+
+A trace sees network latency > 1s when passing between any two services:  
+`{ parent.service.name != service.name } | max(parent.duration - duration) > 1s`
+
+### Shorthand
+
+The following rules are proposed to make using TraceQL easy to use for simple cases.
+
+1. A bare condition is wrapped in `{}`  
+  `http.status = 200` -> `{ http.status = 200 }`
+
+2. Multiple bare conditions are individually wrapped in `{}`    
```

Comment:
Like the idea of shorthand but I also think that shorthand can become source of confusion.

We have `&&` in field expressions where we can write a query like `{ http.status >= 200 && http.status < 300 }`, and `&&` is also being used to write a query like `{ namespace = "prod" } && { http.status = 200 }`

let's say I write two queries
1. `{ namespace = "prod" } && { http.status = 200 }`
2. `{ namespace = "prod" &&  http.status = 200 }`

are these two queries same?

Reply:
they are not the same. (1) selects traces where _some_ spans match 1st and 2nd condition independently. (2) selects traces where both conditions hold on the same spans.

Reply:
I see, thanks.

In that case, I believe suggested shorthand 2 with multiple bare conditions might lead to confusion / unexpected results.

If a user writes `http.status = 200 && namespace = "prod"`, they might expect it return results for `{ http.status = 200 && namespace = "prod" }` (whole query gets wrapped into `{}`, like how it happend for shorthand 1)

But it will return results for `{ http.status = 200 } && { namespace = "prod" }` (wrapping each part into `{}`, as sugessted into shorthand 2)

One way to avoid this is to always wrap whole query into `{}` (only keep shorthand 1) for both cases.

Reply:
I very flexible on the shorthand idea. I mainly wanted the user to type something very quick, easy and intuitive and get an meaningful response. Thanks for this feedback!